_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,421 @@
+R8 in 2001.  The HiPE native compiler is written in Erlang and
+translates the BEAM code to native code without the help of a C
+compiler, therefore avoiding many of the problems that BEAM/C ran into.
+
+The HiPE native compiler can often speed up sequential code by a
+factor of two or three compared to interpreted BEAM code.  We hoped
+that would speed up real-world huge application systems.
+Unfortunately, projects within Ericsson that tried HiPE found that it
+did not improve performance.
+
+Why is that?
+
+The main reason is probably that most huge Erlang applications
+don't contain enough sequential code that HiPE could optimize.  The
+runtime of those systems is typically dominated by some combination of
+message passing, calls to the [ETS BIFs][ets], and garbage collection, none
+of which HiPE can optimize.
+
+[ets]: http://erlang.org/doc/man/ets.html
+
+Another reason could be that big systems typically have many small
+modules.  The HiPE native compiler (in common with the Erlang
+compiler) cannot optimize code across module boundaries, thus being
+unable to do much type-based optimizations.
+
+Also, for most big systems, compiling all Erlang modules to native
+code would lead to impractically long build times and the resulting
+code would consume too much memory.  There is a small overhead of
+switching from native code to interpreted BEAM and vice versa.  It is
+a non-trivial task to figure out which modules that would gain from
+being compiled to native code, and at the same time avoiding an
+excessive amount of context switches between native and interpreted
+code.
+
+Because none of the Ericsson Erlang projects used the HiPE native
+compiler, the OTP team could only afford to spend a limited amount of
+time maintaining HiPE. Therefore, the documentation for HiPE includes
+this note:
+
+> HiPE and execution of HiPE compiled code only have limited support
+  by the OTP team at Ericsson. The OTP team only does limited
+  maintenance of HiPE and does not actively develop HiPE. HiPE is
+  mainly supported by the HiPE team at Uppsala University.
+
+### Other Outcomes the HiPE Project
+
+I think it is fair to say that Erlang/OTP would look very different
+today if it hasn't been for the HiPE project.  Here are the major
+contributions from the HiPE project to OTP:
+
+* A new [staged tag scheme][newtags] in OTP R7.  The new tag scheme
+  allowed the Erlang system to address the full 4GB address space (the
+  previous tag scheme only supported addressing the lower 1 GB).
+  Surprisingly, the new tag scheme also improved performance.
+
+* The [Core Erlang][core] intermediate representation used in the
+  Erlang compiler to this day.  For more information, see
+  [An introduction to Core Erlang][core_introduction] and
+  [Core Erlang by Example][core_by_example].
+
+* [Dialyzer][dialyzer] (DIscrepancy AnaLYZer for ERlang programs),
+  started out as a type analysis pass for the HiPE native compiler,
```

Comment:
```suggestion
  started as a type analysis pass for the HiPE native compiler,
```

Reply:
I will keep "started out" as it is closer to my intended meaning than "started". See https://dictionary.cambridge.org/dictionary/english/start-out.

Reply:
ok!